Road to 💯 - Tests

Here's an abstract list of things the cli does, and how well they're tested (via unit tests)

### Extensions: 
- [x] MySQL
- [x] Nginx (#523, #542)
- [x] Systemd (#544)

### Commands: 
- [x] buster
- [x] install
- [x] log (#545)
- [x] ls
- [x] migrate
- [x] restart
- [x] run
- [x] setup (#559)
- [x] start (#560)
- [x] stop (#582)
- [x] uninstall (#585)
- [x] update 
- [x] version

### Tasks:
- [x] ensure-structure
- [x] linux
- [x] migrate
- [ ] yarn-install (#586)

### Utils:
- [x] config
- [x] find-extensions
- [x] needed-migrations
- [x] resolve-version
- [x] update-check
- [x] use-ghost-user
- [x] version-from-zip
- [ ] yarn (#586)

### Misc:
- [x] Remove spec files from coverage reports (#543)
- [x] bootstrap

When I run the tests locally, there's always one test that fails, but it's fine in Travis:

 ```
 1) Unit: Extensions > Nginx
       migrations hook
         [before 1.2.0]
           Doesn't skip if acme.sh exists:

      AssertionError: expected true to be false
      + expected - actual

      -true
      +false

      at Context.<anonymous> (extensions/nginx/test/extension-spec.js:73:51)
```

--- a/extensions/systemd/test/systemd-spec.js
+++ b/extensions/systemd/test/systemd-spec.js
@@ -61,6 +61,17 @@ describe('Unit: Systemd > Process Manager', function () {
                 expect(error).to.be.instanceOf(errors.ProcessError);
             });
         });
+
+        it('Errors when starting failed', function () {
+            ext.ensureStarted = sinon.stub().rejects(new errors.CliError('Wasn\'t started'));
+            ext.start().then(() => {
+                expect(false, 'Promise should have rejected').to.be.true;
+            }).catch((error) => {
+                expect(error).to.be.ok;
+                expect(ext.ensureStarted.calledOnce).to.be.true;
+                expect(error).to.be.instanceOf(errors.CliError);
+            });
+        });
     });
 
     describe('Stop Hook', function () {
@@ -133,6 +144,17 @@ describe('Unit: Systemd > Process Manager', function () {
                 expect(error).to.be.ok;
                 expect(ui.sudo.calledOnce).to.be.true;
                 expect(error).to.be.instanceOf(errors.ProcessError);
+            });
+        });
+
+        it('Errors when starting failed', function () {
+            ext.ensureStarted = sinon.stub().rejects(new errors.CliError('Wasn\'t started'));
+            ext.restart().then(() => {
+                expect(false, 'Promise should have rejected').to.be.true;
+            }).catch((error) => {
+                expect(error).to.be.ok;
+                expect(ext.ensureStarted.calledOnce).to.be.true;
+                expect(error).to.be.instanceOf(errors.CliError);
             });
         });
     });

--- a/lib/commands/doctor/index.js
+++ b/lib/commands/doctor/index.js
@@ -14,7 +14,7 @@ class DoctorCommand extends Command {
 
         if (!checksToRun.length) {
             if (!argv.quiet) {
-                const additional = argv.categories && argv.categories.length ? ` for categories '${argv.categories.join(', ')}'` : '';
+                const additional = argv.categories && argv.categories.length ? ` for categories "${argv.categories.join(', ')}"` : '';
                 this.ui.log(`No checks found to run${additional}.`);
             }
 

--- a/lib/errors.js
+++ b/lib/errors.js
@@ -97,6 +97,7 @@ class CliError extends Error {
             if (this.err && this.err.message) {
                 output += `${chalk.green('Original Error Message:')}\n`
                 output += `${chalk.gray('Message:')} ${this.err.message}\n`
+                /* istanbul ignore next */
                 if (this.err.stack) {
                     output += `${chalk.gray('Stack:')} ${this.err.stack}\n`
                 }

--- a/lib/instance.js
+++ b/lib/instance.js
@@ -261,6 +261,7 @@ class Instance {
                     return this.template(contents, descriptor, file, dir);
                 }
 
+                /* istanbul ignore else */
                 if (choice === 'edit') {
                     return this.ui.prompt({
                         type: 'editor',

--- a/lib/process-manager.js
+++ b/lib/process-manager.js
@@ -47,6 +47,7 @@ class ProcessManager {
         return this.stop(cwd, env).then(() => this.start(cwd, env));
     }
 
+    /* istanbul ignore next */
     success() {
         // Base implementation - noop
     }
@@ -57,6 +58,7 @@ class ProcessManager {
         throw error;
     }
 
+    /* istanbul ignore next */
     isRunning() {
         // Base Implementation
     }

--- a/lib/ui/index.js
+++ b/lib/ui/index.js
@@ -190,6 +190,7 @@ class UI {
         cp.stderr.on('data', (data) => {
             const lines = data.toString('utf8').split('\n');
 
+            /* istanbul ignore else */
             // If sudo has prompted for a password, then it will be the last line output
             if (lines[lines.length - 1] === prompt) {
                 this.prompt({

--- a/lib/utils/local-process.js
+++ b/lib/utils/local-process.js
@@ -65,6 +65,7 @@ Please ensure the content folder has correct permissions and try again.`));
                     return reject(new errors.GhostError(msg.error));
                 }
 
+                /* istanbul ignore else */
                 if (msg.started) {
                     cp.disconnect();
                     cp.unref();
@@ -123,6 +124,7 @@ Please ensure the content folder has correct permissions and try again.`));
      * @public
      */
     success() {
+        /* istanbul ignore else */
         if (process.send) {process.send({started: true});}
     }
 
@@ -136,6 +138,7 @@ Please ensure the content folder has correct permissions and try again.`));
      * @public
      */
     error(error) {
+        /* istanbul ignore else */
         if (process.send) {process.send({error: true, message: error.message});}
     }
 

--- a/test/fixtures/TestExtension/index.js
+++ b/test/fixtures/TestExtension/index.js
@@ -1,0 +1,9 @@
+'use strict'
+
+const Extension = require('../../../lib/extension');
+
+class TestExtension extends Extension {
+
+}
+
+module.exports = TestExtension;

--- a/test/unit/command-spec.js
+++ b/test/unit/command-spec.js
@@ -211,6 +211,31 @@ describe('Unit: Command', function () {
             }
         });
 
+        it('doesn\'t check for root when command allows root', function () {
+            const checkRootUserStub = sandbox.stub().throws('let them be freeee');
+            class ShortCircuit {
+                constructor() {
+                    throw new Error('you shall not pass');
+                }
+            }
+            const Command = proxyquire(modulePath, {
+                './utils/check-root-user': checkRootUserStub,
+                './ui': ShortCircuit
+            });
+
+            const TestCommand = class extends Command {};
+            TestCommand.global = true;
+            TestCommand.allowRoot = true;
+
+            try {
+                TestCommand._run('test', {});
+            } catch (e) {
+                expect(e).to.exist;
+                expect(e.message).to.equal('you shall not pass');
+                expect(checkRootUserStub.calledOnce).to.be.false;
+            }
+        });
+
         it('loads system and ui dependencies, calls run method', function () {
             const uiStub = sandbox.stub().returns({ui: true});
             const setEnvironmentStub = sandbox.stub();

--- a/test/unit/commands/config-spec.js
+++ b/test/unit/commands/config-spec.js
@@ -243,6 +243,30 @@ describe('Unit: Command > Config', function () {
             expect(result.find(prompt => prompt.name === 'dbpass')).to.be.ok;
             expect(result.find(prompt => prompt.name === 'dbname')).to.not.be.ok;
         });
+
+        it('dbname prompt generates proper default filename', function () {
+            const argv = {
+                url: 'http://localhost.com',
+                db: 'mysql',
+                dbhost: 'localhost',
+                dbuser: 'ghost_rnd',
+                dbpass: 'good_pass'
+            };
+            const getStub = sinon.stub();
+            const ctx = {
+                system: {},
+                instance: {config: {get: getStub}}
+            };
+
+            config.getConfigPrompts.call(ctx, argv);
+            expect(getStub.calledOnce).to.be.true;
+            expect(getStub.getCall(0).args[1]).to.match(/_prod/);
+
+            ctx.system.development = true;
+            config.getConfigPrompts.call(ctx, argv);
+            expect(getStub.calledTwice).to.be.true;
+            expect(getStub.getCall(1).args[1]).to.match(/_dev/);
+        });
     });
 
     describe('run', function () {

--- a/test/unit/commands/doctor/command-spec.js
+++ b/test/unit/commands/doctor/command-spec.js
@@ -17,7 +17,22 @@ describe('Unit: Commands > Doctor', function () {
         return instance.run({}).then(() => {
             expect(listrStub.called).to.be.false;
             expect(logStub.calledOnce).to.be.true;
-            expect(logStub.args[0][0]).to.match(/No checks found to run/);
+            expect(logStub.args[0][0]).to.equal('No checks found to run.');
+        });
+    });
+
+    it('doesn\'t do anything if there are no checks to run (with log + specific categories)', function () {
+        const listrStub = sinon.stub().resolves();
+        const logStub = sinon.stub();
+        const DoctorCommand = proxyquire(modulePath, {
+            './checks': []
+        });
+        const instance = new DoctorCommand({listr: listrStub, log: logStub}, {});
+
+        return instance.run({categories: ['testing', 'validity']}).then(() => {
+            expect(listrStub.called).to.be.false;
+            expect(logStub.calledOnce).to.be.true;
+            expect(logStub.args[0][0]).to.equal('No checks found to run for categories "testing, validity".');
         });
     });
 
@@ -104,7 +119,7 @@ describe('Unit: Commands > Doctor', function () {
         const instance = new DoctorCommand(ui, system);
 
         return instance.run({
-            skipInstanceCheck: true,
+            skipInstanceCheck: false,
             local: true,
             argv: true,
             categories: ['install'],
@@ -117,7 +132,7 @@ describe('Unit: Commands > Doctor', function () {
             expect(ui.listr.args[0][0]).to.deep.equal([{category: ['install']}]);
             const context = ui.listr.args[0][1];
             expect(context.argv).to.deep.equal({
-                skipInstanceCheck: true,
+                skipInstanceCheck: false,
                 local: true,
                 argv: true,
                 categories: ['install'],

--- a/test/unit/commands/log-spec.js
+++ b/test/unit/commands/log-spec.js
@@ -141,6 +141,25 @@ describe('Unit: Commands > Log', function () {
             });
         });
 
+        it('Reads error log when requested', function () {
+            stubs.es.throws(new Error('SHORT_CIRCUIT'));
+            const ext = proxyLog({fs: {existsSync: stubs.es}});
+            ext.system = defaultSystem;
+
+            try {
+                ext.run({name: 'ghost_org', error: true})
+                expect(false, 'existsSync should have thrown').to.be.true;
+            } catch (error) {
+                const fileName = 'https___dev_ghost_org_dev.error.log'
+                const expectedFilePath = `/var/www/ghost/content/logs/${fileName}`;
+                expect(error).to.be.ok;
+                expect(error.message).to.equal('SHORT_CIRCUIT');
+
+                expect(stubs.es.calledOnce).to.be.true;
+                expect(stubs.es.getCall(0).args[0]).to.equal(expectedFilePath);
+            }
+        });
+
         it('Resolves when log file doesn\'t exist', function () {
             stubs.es.returns(false);
             const ext = proxyLog({fs: {existsSync: stubs.es}});

--- a/test/unit/errors-spec.js
+++ b/test/unit/errors-spec.js
@@ -47,6 +47,21 @@ describe('Unit: Errors', function () {
             expect(errorOutput).to.match(/Help: some help message/);
         });
 
+        it('logs suggestion if suggestion option is passed', function () {
+            const errorWithHelp = new errors.CliError({
+                message: 'some error',
+                help: 'some help message',
+                suggestion: 'run ghost doctor of course'
+            });
+
+            const errorOutput = stripAnsi(errorWithHelp.toString()).trim().split(/\n/);
+            const expected = ['Message: some error',
+                'Help: some help message',
+                'Suggestion: run ghost doctor of course'
+            ];
+            expect(errorOutput).to.deep.equal(expected);
+        });
+
         it('logs original error message and stack trace if verbose is set', function () {
             const originalError = new Error('something aweful happened here');
             originalError.response = 'very long response';

--- a/test/unit/extension-spec.js
+++ b/test/unit/extension-spec.js
@@ -127,4 +127,22 @@ describe('Unit: Extension', function () {
         expect(logStub.calledOnce).to.be.true;
         expect(logStub.args[0][0]).to.match(/not a valid Extension subclass/);
     });
+
+    it('returns an extension subclass if everything works out', function () {
+        const path = require('path');
+        const Extension = require(modulePath);
+        const testExtPath = '../fixtures/TestExtension';
+        const TestExt = require(testExtPath);
+
+        const extensionInstance = Extension.getInstance({}, {}, {
+            pkg: {
+                name: 'ghost-cli-test-extension',
+                main: 'index.js'
+            },
+            dir: path.resolve(__dirname, testExtPath)
+        });
+
+        expect(extensionInstance).to.be.ok;
+        expect(extensionInstance instanceof TestExt).to.be.true;
+    });
 });

--- a/test/unit/tasks/yarn-install-spec.js
+++ b/test/unit/tasks/yarn-install-spec.js
@@ -41,6 +41,27 @@ describe('Unit: Tasks > yarn-install', function () {
         });
     });
 
+    it('base function can take zip file', function () {
+        const decompressStub = sinon.stub().resolves();
+        const listrStub = sinon.stub().resolves();
+        const yarnInstall = proxyquire(modulePath, {
+            decompress: decompressStub
+        });
+
+        return yarnInstall({listr: listrStub}, 'test.zip').then(() => {
+            const ctx = {installPath: '/var/www/ghost'};
+            expect(listrStub.calledOnce).to.be.true;
+
+            const tasks = listrStub.getCall(0).args[0];
+            expect(tasks).to.have.length(2);
+
+            tasks[0].task(ctx);
+
+            expect(decompressStub.called).to.be.true;
+            expect(decompressStub.calledWithExactly('test.zip','/var/www/ghost')).to.be.true;
+        });
+    });
+
     describe('dist subtask', function () {
         it('rejects if yarn util returns invalid json', function () {
             const yarnStub = sinon.stub().resolves({stdout: 'not json'});

--- a/test/unit/ui/index-spec.js
+++ b/test/unit/ui/index-spec.js
@@ -269,6 +269,24 @@ describe('Unit: UI', function () {
         stderr.emit('data', '#node-sudo-passwd#');
     });
 
+    it('#sudo can handle defaults', function () {
+        const shell = Promise.resolve();
+        shell.stderr = {on: () => true};
+
+        const shellStub = sinon.stub().returns(shell);
+
+        const UI = proxyquire(modulePath, {execa: {shell: shellStub}});
+        const ui = new UI();
+
+        ui.log = () => true;
+        ui.prompt = sinon.stub();
+
+        return ui.sudo('echo').then(() => {
+            expect(shellStub.calledOnce).to.be.true;
+            expect(shellStub.getCall(0).args[0]).to.match(/#'[ ]{2}echo/);
+        });
+    });
+
     describe('#noSpin', function () {
         let ui;
 

--- a/test/unit/utils/find-extensions-spec.js
+++ b/test/unit/utils/find-extensions-spec.js
@@ -1,8 +1,62 @@
 'use strict';
 
+const expect = require('chai').expect;
+const sinon = require('sinon');
+const proxyquire = require('proxyquire').noCallThru();
+
 const modulePath = '../../../lib/utils/find-extensions';
 
-// TODO: remove line once tests are implemented
-require(modulePath);
+const localExtensions = [
+    'linux',
+    'nginx',
+    'mysql',
+    'systemd'
+];
 
-describe.skip('Unit: Utils > find-extensions', function () {});
+describe('Unit: Utils > find-extensions', function () {
+    let findExtensions, findStub;
+
+    beforeEach(() => {
+        findStub = sinon.stub().returns([
+            {
+                pkg: {name: 'test'}
+            }, {
+                pkg: {
+                    'ghost-cli': {name: 'rest'}
+                }
+            }, {
+                pkg: {}
+            }
+        ]);
+
+        findExtensions = proxyquire(modulePath, {
+            'find-plugins': findStub,
+            execa: {shellSync: () => ({stdout: '.'})}
+        });
+    });
+
+    it('calls find-plugins with proper args', function () {
+        findExtensions();
+        expect(findStub.calledOnce).to.be.true;
+        const args = findStub.getCall(0).args[0];
+
+        const expected = {
+            keyword: 'ghost-cli-extension',
+            configName: 'ghost-cli',
+            scanAllDirs: true,
+            dir: '.',
+            sort: true
+        };
+
+        const extensions = args.include.map((ext) => ext.split('extensions/')[1]);
+        delete args.include;
+        expect(extensions).to.deep.equal(localExtensions);
+        expect(args).to.deep.equal(expected);
+    });
+
+    it('generates proper extension names', function () {
+        const names = findExtensions().map((ext) => ext.name);
+        const expectedNames = ['test', 'rest', undefined];
+        expect(names).to.deep.equal(expectedNames);
+    });
+});

--- a/test/unit/utils/get-os-spec.js
+++ b/test/unit/utils/get-os-spec.js
@@ -19,7 +19,7 @@ describe('Unit: Utils > getOS', function () {
         sandbox.restore();
     });
 
-    it('and returns correct Linux OS', function () {
+    it('returns correct Linux OS', function () {
         platformStub.returns('linux');
         execaStub.withArgs('lsb_release -i -s').returns({stdout: 'Ubuntu'});
         execaStub.withArgs('lsb_release -r -s').returns({stdout: '16'});
@@ -30,7 +30,7 @@ describe('Unit: Utils > getOS', function () {
         expect(execaStub.calledTwice).to.be.true;
     });
 
-    it('and returns correct mac OS', function () {
+    it('returns correct mac OS', function () {
         platformStub.returns('darwin');
         execaStub.withArgs('sw_vers -productName').returns({stdout: 'Mac OS X'});
         execaStub.withArgs('sw_vers -productVersion').returns({stdout: '10.13.3'});
@@ -41,7 +41,7 @@ describe('Unit: Utils > getOS', function () {
         expect(execaStub.calledTwice).to.be.true;
     });
 
-    it('and returns correct Windows OS', function () {
+    it('returns correct Windows OS', function () {
         platformStub.returns('win32');
         execaStub.withArgs('ver').returns({stdout: 'Microsoft Windows XP [Version 5.1.2600]'});
 
@@ -51,7 +51,7 @@ describe('Unit: Utils > getOS', function () {
         expect(execaStub.calledOnce).to.be.true;
     });
 
-    it('and returns default os.platform if OS is not Mac, Linux, or Windows', function () {
+    it('returns default os.platform if OS is not Mac, Linux, or Windows', function () {
         platformStub.returns('freebsd');
         versionStub.returns('1.0.0')
         const osResult = getOS({
@@ -62,5 +62,35 @@ describe('Unit: Utils > getOS', function () {
         expect(osResult.os).to.equal('freebsd');
         expect(osResult.version).to.equal('1.0.0');
         expect(execaStub.calledOnce).to.be.false;
+    });
+
+    it('returns default os when syscall fails', function () {
+        platformStub.returns('linux');
+        versionStub.returns('1.0.0');
+        execaStub.throws();
+
+        let osResult = getOS({
+            linux: true
+        });
+
+        expect(osResult.os).to.equal('linux');
+        expect(osResult.version).to.equal('1.0.0');
+        expect(execaStub.calledOnce).to.be.true;
+
+        osResult = getOS({
+            macos: true
+        });
+
+        expect(osResult.os).to.equal('linux');
+        expect(osResult.version).to.equal('1.0.0');
+        expect(execaStub.calledTwice).to.be.true;
+
+        osResult = getOS({
+            windows: true
+        });
+
+        expect(osResult.os).to.equal('linux');
+        expect(osResult.version).to.equal('1.0.0');
+        expect(execaStub.calledThrice).to.be.true;
     });
 });

--- a/test/unit/utils/resolve-version-spec.js
+++ b/test/unit/utils/resolve-version-spec.js
@@ -35,6 +35,18 @@ describe('Unit: resolveVersion', function () {
         });
     });
 
+    it('defaults versions to none', function () {
+        stubYarn('{}');
+
+        return resolveVersion().then(function () {
+            throw new Error('Version finder should not have resolved.');
+        }).catch(function (error) {
+            expect(error).to.be.an.instanceOf(Error);
+            expect(error.message).to.equal('No valid versions found.');
+        });
+    });
+
+
     it('rejects if no versions are found', function () {
         stubYarn('{"data": []}');
 

--- a/test/unit/utils/yarn-spec.js
+++ b/test/unit/utils/yarn-spec.js
@@ -2,6 +2,9 @@
 const expect = require('chai').expect;
 const proxyquire = require('proxyquire');
 const sinon = require('sinon');
+const ProcessError = require('../../../lib/errors').ProcessError;
+
+const modulePath = '../../../lib/utils/yarn';
 
 let yarn;
 
@@ -13,8 +16,8 @@ describe('Unit: yarn', function () {
         currentEnv = process.env;
         process.env = {};
 
-        execa = sinon.stub().returns(new Promise((resolve) => {resolve();}));
-        yarn = proxyquire('../../../lib/utils/yarn', {
+        execa = sinon.stub().resolves();
+        yarn = proxyquire(modulePath, {
             execa: execa
         });
     });
@@ -68,6 +71,123 @@ describe('Unit: yarn', function () {
             expect(execa.args[0][2].env).to.be.an('object');
             expect(execa.args[0][2].env.TESTENV).to.equal('override');
             expect(process.env.TESTENV).to.equal('test');
+        });
+    });
+
+    it('fails gracefully when yarn fails', function () {
+        execa.rejects(new Error('YARN_TO_FAST'));
+        yarn = proxyquire(modulePath, {execa: execa});
+
+        return yarn().then(() => {
+            expect(false, 'Promise should have rejected').to.be.true;
+        }).catch((error) => {
+            expect(execa.calledOnce).to.be.true;
+            expect(error).to.be.ok;
+            expect(error).to.be.instanceOf(ProcessError);
+        });
+    })
+
+    describe('can return observables', function () {
+        let stubs;
+        beforeEach(function () {
+            stubs = {
+                stdout: sinon.stub(),
+                observer: {
+                    next: sinon.stub(),
+                    complete: sinon.stub(),
+                    error: sinon.stub()
+                }
+            };
+
+            class TestClass {
+                constructor(fn) {
+                    stubs.proxy = {fn: fn};
+                    return (new Promise((resolve,reject) => {
+                        stubs.proxy.resolve = resolve;
+                        stubs.proxy.reject = reject;
+                    }));
+                }
+            }
+
+            const execaPromise = new Promise((resolve, reject) => {
+                stubs._execa = {
+                    resolve: resolve,
+                    reject: reject
+                };
+            });
+            execaPromise.stdout = {
+                setEncoding: () => true,
+                on: stubs.stdout
+            };
+            execa.returns(execaPromise);
+
+            yarn = proxyquire(modulePath, {
+                execa: execa,
+                rxjs: {Observable: TestClass}
+            });
+        })
+
+        it('ends properly', function () {
+            const res = yarn([], {observe: true});
+            expect(stubs.proxy).to.be.an('Object');
+
+            stubs.proxy.fn(stubs.observer);
+            stubs._execa.resolve();
+            res.then(() => {
+                expect(stubs.observer.complete.calledOnce).to.be.true;
+            });
+
+            stubs.proxy.resolve();
+            return res;
+        });
+
+        it('ends properly (error)', function () {
+            const res = yarn([], {observe: true});
+            expect(stubs.proxy).to.be.an('Object');
+            stubs.proxy.fn(stubs.observer);
+            stubs._execa.reject('test');
+
+            res.then(() => {
+                expect(stubs.observer.complete.called).to.be.false;
+                expect(stubs.observer.error.calledOnce).to.be.true;
+                expect(stubs.observer.error.getCall(0).args[0]).to.equal('test');
+            });
+
+            stubs.proxy.resolve();
+            return res;
+        });
+
+        it('proxies data through', function () {
+            yarn([], {observe: true});
+            expect(stubs.proxy).to.be.an('Object');
+            stubs.proxy.fn(stubs.observer);
+            expect(stubs.stdout.calledOnce).to.be.true;
+            expect(stubs.stdout.getCall(0).args[0]).to.equal('data');
+
+            const onFn = stubs.stdout.getCall(0).args[1];
+            onFn('\n\n\n');
+            onFn('test\n');
+            onFn('\nbest\n');
+            onFn('run');
+
+            const next = stubs.observer.next
+            expect(next.callCount).to.equal(4);
+            expect(next.getCall(0).args[0]).to.equal('\n\n');
+            expect(next.getCall(1).args[0]).to.equal('test');
+            expect(next.getCall(2).args[0]).to.equal('\nbest');
+            expect(next.getCall(3).args[0]).to.equal('run');
+        });
+
+        it('cleans up observer error', function () {
+            const res = yarn([], {observe: true});
+            expect(stubs.proxy).to.be.an('Object');
+            stubs.proxy.reject();
+            return res.then(() => {
+                expect(false, 'Promise should have rejected').to.be.true;
+            }).catch((e) => {
+                expect(e).to.be.ok;
+                expect(e).to.be.instanceOf(ProcessError);
+            })
         });
     });
 });

